hikari/impl/stateful_guild_chunker.py

```diff
@@ -23,65 +23,298 @@
+
+    average_chunk_size: typing.Optional[int] = attr.ib(default=None)
+    chunk_count: typing.Optional[int] = attr.ib(default=None)
+    guild_id: snowflakes.Snowflake = attr.ib()
+    last_received: typing.Optional[datetime.datetime] = attr.ib(default=None)
+    _mono_last_received: typing.Optional[int] = attr.ib(default=None)
+    missing_chunk_indexes: typing.Optional[typing.MutableSequence[int]] = attr.ib(default=None)
+    nonce: str = attr.ib()
+    not_found_ids: typing.MutableSequence[snowflakes.Snowflake] = attr.ib(factory=list)
+
+    def __copy__(self) -> _TrackedRequests:
+        chunks = attr_extensions.copy_attrs(self)
+        chunks.missing_chunk_indexes = (
+            list(self.missing_chunk_indexes) if self.missing_chunk_indexes is not None else None
+        )
+        chunks.not_found_ids = list(self.not_found_ids)
+        return chunks
+
+    @property
+    def is_complete(self) -> bool:
+        if self.received_chunks == self.chunk_count:
+            return True
+
+        return self._mono_last_received is not None and time.monotonic_ns() - self._mono_last_received > EXPIRY_TIME
+
+    @property
+    def received_chunks(self) -> int:
+        if self.chunk_count is None or self.missing_chunk_indexes is None:
+            return 0
+
+        return self.chunk_count - len(self.missing_chunk_indexes)
+
+    def update(self, event: shard_events.MemberChunkEvent) -> None:
+        if self.missing_chunk_indexes is None:
+            self.average_chunk_size = len(event.members)
+            self.chunk_count = event.chunk_count
+            self.missing_chunk_indexes = list(range(event.chunk_count))
+
+        self.not_found_ids.extend(event.not_found)
+        self.missing_chunk_indexes.remove(event.chunk_index)
+        self._mono_last_received = time.monotonic_ns()
+        self.last_received = date.utc_datetime()
 
 
 class StatefulGuildChunkerImpl(chunker.GuildChunker):
-    """Guild chunker implementation."""
+    """Guild chunker implementation.
 
-    __slots__: typing.Sequence[str] = ("_app", "_presences", "_queues", "_chunkers")
+    Parameters
+    ----------
+    app: hikari.traits.BotAware
+        The object of the bot aware app this is bound to.
+    limit: builtins.int
+        The maximum amount of requests that this chunker should store information
+        about for each shard.
+    """
 
-    def __init__(self, app: traits.ShardAware, intents: typing.Optional[intents_.Intents]):
+    __slots__: typing.Sequence[str] = ("_app", "_tracked")
+
+    def __init__(self, app: traits.BotAware, limit: int = 200) -> None:
         self._app = app
-        self._presences: bool = intents is None or bool(intents & intents_.Intents.GUILD_PRESENCES)
-        self._queues: typing.Dict[int, typing.List[int]] = {}
-        self._chunkers: typing.Dict[int, asyncio.Task[None]] = {}
+        self._tracked: typing.MutableMapping[int, typing.Dict[str, _TrackedRequests]] = mapping.CMRIMutableMapping(
```

Comment:
Can we just use theCRMIMapping type as the hint

Reply:
Is there any actual reason to doing that when for all intensive purposes the CRMIMapping is still a normal MutableMapping interface wise